### DISCUSSION TITLE
How to know the current process when create database instance

### DISCUSSION THREAD

**Discussion by dim0147** [simolus3/drift#1344] [Q&A]
Hi, the title may confuse, here is my case, when first time open the app my database will connect and create tables, after finish created tables it will create some default  data into some tables, I want to know how I can obtain those process to display on my UI(Widget) follow steps like 'Connecting to db...' > 'Create tables...' > 'Populate data first time open app...' thank you.

**Comment by simolus3** [comment]
You can just report the state yourself:

```dart
enum OpeningState {
  closed,
  openingStarted,
  creatingTables,
  creatingInitialData,
  opened
}

class MyDatabase extends _$MyDatabase {
  final BehaviorSubject<OpeningState> state = BehaviorSubject.seeded(OpeningState.closed);

  MigrationStrategy get migration {
    state.add(OpeningState.openingStarted);
    return MigrationStrategy(
      onCreate: (m) {
        state.add(OpeningState.creatingTables);
        await m.createAll();
        state.add(OpeningState.creatingInitialData);
        await yourFunctionToInsertData();
      },
      beforeOpen: (details) => state.add(OpeningState.opened),
    );
  }
}
```

But in practice, I hope opening the database and setting everything up should be so fast that it's not noticeable. So I'm not sure why you should show a UI progress for that, did you encounter this taking a long time?

**Reply by dim0147** [reply]
Thank you for your reply, I don't encounter those processes taking a long time since I have work on the remote databases before and it takes a little time waiting for connecting, but this is local storage so I think no need to worry about this. Can I ask you this, currently my small shop needs to store order detail into this local storage around 500 records per day, in few months it may become 20k or 30k records, is moor will suitable for handling large records? I wonder when the app starts it will load the **db.sql** file right? So if this file contains 20k records, will it be taking a long time to load this file (5 or 6 seconds I think), if so,  should I avoid using local storage and using the remote database? Thank you

**Reply by simolus3** [reply]
> Thank you for your reply, I don't encounter those processes taking a long time since I have work on the remote databases before and it takes a little time waiting for connecting, but this is local storage so I think no need to worry about this.

Exactly right - there's no expensive connection here, "connecting" to a file-based database should be a matter of milliseconds.

sqlite3 supports large databases, its not all read at once. Moor is just wrapping sqlite3, so the same ideas apply. If you're inserting 30k records _at once_, there might be a bit of an delay but if you do it incrementally I don't see a problem.

**Reply by dim0147** [reply]
Thank you, this really big help from me, I was thinking to buy a small **raspberry pi** to host the database but seems I will use the local storage for this. Thank you

**Reply by simolus3** [reply]
Just out of interest, what are you building? When you have a database server that's typically the central source of truth for all users, a local database on the client side still needs to fetch data from somewhere. In other words, how do the records end up in the local database?

**Reply by dim0147** [reply]
I'm building the app to manage a coffee shop, it only needs simple functionality like show coffee for the customer to select and then places an order, that's it, no need for a remote server because sometimes the wifi won't work (my place don't have good network condition). All data is store in the device.

**Reply by simolus3** [reply]
Ah I see, so you only have a single relevant device then. That sounds cool!
